src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2749,13 +2813,560 @@ def query(self, filters=None, options=None):
+        options = options or {}
+        recursive_op = (
+            (lambda x, y: f'{x}/'.startswith(f'{y}/')) if options.get('recursive') else (lambda x, y: x == y)
+        )
+        key_loaded = options.get('key_loaded', True)
+        db_results = self.encrypted_roots_query_db(
+            [
+                ['name', '=', name], *(['name', '^', f'{name}/'] if options.get('recursive') else []),
+                *(options.get('db_filters', []))
+            ], True
+        ) if options.get('load_db') else {}
+
+        def normalize(ds):
+            passphrase = ZFSKeyFormat(ds['key_format']['value']) == ZFSKeyFormat.PASSPHRASE
+            key = db_results.get(ds['name']) if not passphrase else None
+            return ds['name'], {'encryption_key': key, **ds} if options.get('full_output') else key
+
+        def check_key(ds):
+            return options.get('all') or (ds['key_loaded'] and key_loaded) or (not ds['key_loaded'] and not key_loaded)
+
+        return dict(map(
+            normalize,
+            filter(
+                lambda d: d['name'] == d['encryption_root'] and d['encrypted'] and recursive_op(
+                    d['name'], name
+                ) and check_key(d),
+                self.query()
+            )
+        ))
+
+    @periodic(86400)
+    @private
+    @job(lock=lambda args: f'sync_encrypted_pool_dataset_keys_{args}')
+    def sync_db_keys(self, job, filters=None):
+        db_datasets = self.encrypted_roots_query_db(filters=filters, decrypt=True)
+        to_remove = []
+        check_key_job = self.middleware.call_sync('zfs.dataset.bulk_process', 'check_key', [
+            (name, {'key': db_datasets[name]}) for name in db_datasets
+        ])
+        check_key_job.wait_sync()
+        if check_key_job.error:
+            self.middleware.logger.error(f'Failed to sync database keys: {check_key_job.error}')
+
+        for dataset, status in zip(db_datasets, check_key_job.result):
+            if not status['result'] or status['error']:
+                to_remove.append(dataset)
+
+        self.middleware.call_sync('datastore.delete', self.dataset_store, [['name', 'in', to_remove]])
+
+    @accepts(Str('id'))
+    @job(lock='dataset_export_keys', pipes=['output'])
+    def export_keys(self, job, id):
+        """
+        Export keys for `id` and it's children which are stored in the system. The exported file is a compressed
+        tarfile containing files named after the encrypted dataset and containing the keys for them.
+
+        Please refer to websocket documentation for downloading the file.
+        """
+        self.middleware.call_sync('pool.dataset._get_instance', id)
+        sync_job = self.middleware.call_sync('pool.dataset.sync_db_keys', [['name', '^', id]])
+        sync_job.wait_sync()
+
+        datasets = self.encrypted_roots_query_db([['name', '^', id]], decrypt=True)
+        tar_path = None
+        try:
+            tar_path = tempfile.mkstemp()[1]
```

Comment:
Why not tempfile.NamedTemporaryFile?

Reply:
That will provide me with an open file handle which i'll have to close. That's not the case with `mkstemp`